core/crypto/kem/hybrid/hybrid.go

```diff
@@ -0,0 +1,341 @@
+	second := sch.second.EncapsulationSeedSize()
+	ret := second
+	if first > second {
+		ret = first
+	}
+	return ret
+}
+
+func (sk *PrivateKey) Scheme() kem.Scheme { return sk.scheme }
+func (pk *PublicKey) Scheme() kem.Scheme  { return pk.scheme }
+
+func (sk *PrivateKey) MarshalBinary() ([]byte, error) {
+	if sk.first == nil || sk.second == nil {
+		return nil, ErrUninitialized
+	}
+	first, err := sk.first.MarshalBinary()
+	if err != nil {
+		return nil, err
+	}
+	second, err := sk.second.MarshalBinary()
+	if err != nil {
+		return nil, err
+	}
+	return append(first, second...), nil
+}
+
+func (sk *PrivateKey) Equal(other kem.PrivateKey) bool {
+	oth, ok := other.(*PrivateKey)
+	if !ok {
+		return false
+	}
+	if sk.first == nil && sk.second == nil && oth.first == nil && oth.second == nil {
+		return true
+	}
+	if sk.first == nil || sk.second == nil || oth.first == nil || oth.second == nil {
+		return false
+	}
+	return sk.first.Equal(oth.first) && sk.second.Equal(oth.second)
+}
+
+func (sk *PrivateKey) Public() kem.PublicKey {
+	return &PublicKey{sk.scheme, sk.first.Public(), sk.second.Public()}
+}
+
+func (pk *PublicKey) Equal(other kem.PublicKey) bool {
+	oth, ok := other.(*PublicKey)
+	if !ok {
+		return false
+	}
+	if pk.first == nil && pk.second == nil && oth.first == nil && oth.second == nil {
+		return true
+	}
+	if pk.first == nil || pk.second == nil || oth.first == nil || oth.second == nil {
+		return false
+	}
+	return pk.first.Equal(oth.first) && pk.second.Equal(oth.second)
+}
+
+func (pk *PublicKey) MarshalBinary() ([]byte, error) {
+	if pk.first == nil || pk.second == nil {
+		return nil, ErrUninitialized
+	}
+	first, err := pk.first.MarshalBinary()
+	if err != nil {
+		return nil, err
+	}
+	second, err := pk.second.MarshalBinary()
+	if err != nil {
+		return nil, err
+	}
+	return append(first, second...), nil
+}
+
+func (sch *Scheme) GenerateKeyPair() (kem.PublicKey, kem.PrivateKey, error) {
+	pk1, sk1, err := sch.first.GenerateKeyPair()
+	if err != nil {
+		return nil, nil, err
+	}
+	pk2, sk2, err := sch.second.GenerateKeyPair()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	return &PublicKey{sch, pk1, pk2}, &PrivateKey{sch, sk1, sk2}, nil
+}
+
+func (sch *Scheme) DeriveKeyPair(seed []byte) (kem.PublicKey, kem.PrivateKey) {
+	if len(seed) != sch.SeedSize() {
+		panic(kem.ErrSeedSize)
+	}
+	h := sha3.NewShake256()
+	_, _ = h.Write(seed)
+	first := make([]byte, sch.first.SeedSize())
+	second := make([]byte, sch.second.SeedSize())
+	_, _ = h.Read(first)
+	_, _ = h.Read(second)
+
+	pk1, sk1 := sch.first.DeriveKeyPair(first)
+	pk2, sk2 := sch.second.DeriveKeyPair(second)
+
+	return &PublicKey{sch, pk1, pk2}, &PrivateKey{sch, sk1, sk2}
+}
+
+func splitPRF(ss1, ss2, ct1, ct2 []byte) []byte {
+	// implement split PRF KEM combiner as:
+	// H(ss1 || ss2 || ct1 || ct2)
+	// in order to retain IND-CCA2 security
+	// as described here:
+	// https://link.springer.com/chapter/10.1007/978-3-319-76578-5_7
```

Comment:
At a first glance the extra hashing in of `ct1, ct2` here seems pointless. Does the paper suggest that it's stronger this way? Seems to me that putting attacker-controlled inputs directly into the combiner here would be riskier than the hash-of-hash product, but I'd be interesting to read what they say.

The paper referenced has a lot of different constructions; we should write explicitly which one we are trying to implement, with a page reference and a unique identifier.
Otherwise the reviewer has to read a 30 page cryptography paper and then try to match an implementation, which may or may not actually match something in the paper, to any of the constructions there. I'm not going to do that today. :-)

Reply:
Let's talk about this offline, since a cryptographer already told me to use H(ss1 || ss2 || ct1 || ct2)

Reply:
We're not hashing `H(ss1 || ss2 || ct1 || ct2)` here. The `ss1` and `ss2` parameters to this function are (more or less) `H(ss_x, ct_x)`, thus `ct_x` (the public keys) are hashed in twice.
The linked-to paper is a very long paper and does not appear to have a quote a cryptographer recommending this nested scheme. It's still unclear which scheme from the paper this function was intended to implement.

Reply:
The current construction as I've written it is perfectly correct but I can see how it's maybe maybe slightly confusing but let me explain. We do NOT care at all what each individual KEM does when we are working on the code inside the KEM combiner... except that we start with the assumption that we are combining two KEMs which both have IND-CCA2 security. The idea here is that if one of the KEMs has an attack whereby it's security is reduced to IND-CPA then in that case this KEM combiner will still retain the IND-CCA2 security. Happy to chat with you about this outside of this pull request.

Reply:
1. It's "perfectly correct" in the sense that it's not harmful iff the hash function has no issues (length extension attacks, etc).
If it has, `H(H(ss1 || ct1) || H(ss2 || ct2))` would be better since bogus inputs to ct1/ct2 would not be able to directly influence the hashing state of the combined value.
In the single-KEM instance, `H(ct || ss)` is probably more robust than `H(ss || ct)`.

2. It's still unclear to me on what page of the referenced paper this scheme is discussed.